src/pybind/mgr/cephadm/module.py

```diff
@@ -1810,11 +1824,16 @@ def _check_daemons(self):
                 continue
 
             # dependencies?
-            if dd.daemon_type == 'grafana':
-                # put running instances at the front of the list
-                grafanas.insert(0, dd)
-            elif dd.daemon_type == 'iscsi':
-                iscsi_daemons.append(dd)
+            if dd.daemon_type in ['grafana', 'iscsi', 'prometheus', 'alertmanager']:
+                cephadm_service = self.cephadm_services.get(dd.daemon_type)
+                v = post_actions.setdefault(dd.daemon_type,
+                                            dict(service=cephadm_service, daemons=[]))
+                if dd.daemon_type == 'grafana':
+                    # put running instances at the front of the list
+                    v['daemons'].insert(0, dd)
+                else:
+                    v['daemons'].append(dd)
+
```

Comment:
Instead of doing this per-service guessing shutdown logic (if cassandra then this, else that), why not simply extending `CephAdmSerivce` with some `post_actions` method? Maybe it's just me that I'm not familiar with cephadm module code, but in general I find really tricky when the code has to do a lot of dictionary looping/crunching/unfolding/... for things that would be far clearer with structured classes.

Reply:
Fixes. it's done in service's own function. We just collect daemons here.